app/models/hca_azul_client.rb

```diff
@@ -0,0 +1,138 @@
+      context = " encountered when requesting '#{path}', attempt ##{current_retry}"
+      log_message = "#{e.message}: #{e.http_body}; #{context}"
+      Rails.logger.error log_message
+      retry_time = retry_count * ApiHelpers::RETRY_INTERVAL
+      sleep(retry_time)
+      # only retry if status code indicates a possible temporary error, and we are under the retry limit and
+      # not calling a method that is blocked from retries
+      if should_retry?(e.http_code) && retry_count < ApiHelpers::MAX_RETRY_COUNT
+        process_api_request(http_method, path, payload: payload, retry_count: current_retry)
+      else
+        # we have reached our retry limit or the response code indicates we should not retry
+        ErrorTracker.report_exception(e, nil, {
+          method: http_method, url: path, payload: payload, retry_count: retry_count
+        })
+        error_message = parse_response_body(e.message)
+        Rails.logger.error "Retry count exceeded when requesting '#{path}' - #{error_message}"
+        raise e
+      end
+    end
+  end
+
+  # take a Hash/JSON object and format as a query string parameter
+  #
+  # * *params*
+  #   - +query_params+ (Hash) => Hash of query parameters
+  #
+  # * *returns*
+  #   - (String) => URL-encoded string version of query parameters
+  def format_hash_as_query_string(query_params)
+    # replace Ruby => assignment operators with JSON standard colons (:)
+    sanitized_params = query_params.to_s.gsub(/=>/, ':')
+    CGI.escape(sanitized_params)
+  end
+
+  # API endpoint bindings
+
+  # get a metadata TSV file for a given HCA project UUID
+  #
+  # * *params*
+  #   - +catalog+ (String) => HCA catalog name, from HCA_CATALOGS
+  #   - +project_id+ (UUID) => HCA project UUID
+  #   - +format+ (string) => manifest file format, from MANIFEST_FORMATS
+  #
+  # * *returns*
+  #   - (Hash) => Hash response including an HTTP status code and a location to download
+  #
+  # * *raises*
+  #   - (ArgumentError) => if catalog is not in HCA_CATALOGS or format is not in MANIFEST_FORMATS
+  def get_project_manifest_link(catalog, project_id, format = 'compact')
+    unless HCA_CATALOGS.include?(catalog)
+      raise ArgumentError, "#{catalog} is not a valid catalog: #{HCA_CATALOGS.join(',')}"
+    end
+    unless MANIFEST_FORMATS.include?(format)
+      raise ArgumentError, "#{format} is not a valid format: #{MANIFEST_FORMATS.join(',')}"
+    end
+
+    path = self.api_root + "/fetch/manifest/files?catalog=#{catalog}"
+    project_filter = { 'projectId' => { 'is' => [project_id] } }
+    filter_query = format_hash_as_query_string(project_filter)
+    path += "&filters=#{filter_query}&format=#{format}"
+    # since manifest files are generated on-demand, keep making requests until the Status code is 302 (Found)
+    # Status 301 means that the manifest is still being generated; if no manifest is ready after 30s, return anyway
+    time_slept = 0
+    manifest_info = process_api_request(:get, path)
```

Comment:
I wonder why Azul generates the manifest files on-demand.  Presumably, their upstream components don't frequently change.

Reply:
This is because the endpoint we are using (`/fetch/manifest/files`) supports a query interface.  For our usage, we're only asking for project-level information by passing the `{ 'projectId' => { 'is' => [project_id] } }` query, but you can provide more detailed queries to further refine the row-level data returned.  Logically, there's no way that Azul can precompute everything, so manifests are generated on-demand.

You can read more about the particular endpoint here: https://service.azul.data.humancellatlas.org/#/Manifests/get_fetch_manifest_files

Reply:
Makes sense, thanks for the pointer!